Snapshot and Dumps redesign (#552)
meilisearch/documentation: guides/advanced_guides/configuration.md

- [Analytics](/guides/advanced_guides/configuration.md#analytics)
- [Payload Limit Size](/guides/advanced_guides/configuration.md#payload-limit-size)
- [Dumps](/guides/advanced_guides/configuration.md#dumps-destination)

>   - [Dumps Destination](/guides/advanced_guides/configuration.md#dumps-destination)

Reviewer: I like this rename a lot

- [Analytics](/guides/advanced_guides/configuration.md#analytics)
- [Payload Limit Size](/guides/advanced_guides/configuration.md#payload-limit-size)

> - [Snapshoting](/guides/advanced_guides/configuration.md#schedule-snapshot-creation):

Reviewer 2: ```suggestion
- [Snapshots](/guides/advanced_guides/configuration.md#schedule-snapshot-creation):
```

  - [Schedule Snapchot Creation](/guides/advanced_guides/configuration.md#schedule-snapshot-creation)
  - [Snapshot Destination](/guides/advanced_guides/configuration.md#snapshot-destination)
  - [Snapshot Interval](/guides/advanced_guides/configuration.md#snapshot-interval)
  - [Import Snapshot](/guides/advanced_guides/configuration.md#import-snapshot)
  - [Ignore Missing Snapshot](/guides/advanced_guides/configuration.md#ignore-missing-snapshot)

>   - [Ignore Snapshot If DB Exists](/guides/advanced_guides/configuration.md#ignore-snapshot-if-db-exists)

Reviewer 2: ```suggestion
  - [Ignore Snapshot if DB Exists](/guides/advanced_guides/configuration.md#ignore-snapshot-if-db-exists)
```

### Schedule Snapshot Creation

**Environment variable**: `MEILI_SCHEDULE_SNAPSHOT`
**CLI option**: `--schedule-snapshot`

> Activates snapshots scheduling.

Reviewer 2: ```suggestion
Activates scheduled snapshots.
```

> The directory path where MeiliSearch will create snapshots.
> 
> If this command is not called, snapshoting is deactivated.

Reviewer: ```suggestion
Activates snapshots scheduling. 

If this command is not added or its value is `false` snapshotting is deactivated.
```

If this command is not called, no snapshot will be imported.

### Ignore Missing Snapshot

**Environment variable**: `MEILI_IGNORE_MISSING_SNAPSHOT`
**CLI option**: `--ignore-missing-snapshot`

> The engine ignores missing snapshots and does not throw an error in this case.

Reviewer: ```suggestion
The engine ignores missing snapshots in the case of a `--snapshot-import`, and does not throw an error in this case.
```

Author: ```suggestion
The engine ignores missing snapshots in the case of a `--import-snapshot`, and does not throw an error in this case.
```

Reviewer: I want to make it clear that this variable has to do with the import snapshot process.

Author: just show you the real option, because it's less natural in reading with this version but "more true" in the usage

Reviewer: Ah yes! Ok I had to read it 200 times before seeing the difference. My bad!

> Requires `--import-snapshot` to be defined.

Reviewer: Does this mean an error is raised if the flag is given but `--import-snapshot` is not?

Author: yep

Reviewer: Is this not considered like `--snapshot-path` as a setting that can be pre-added as an ENV variable but will only work once we add `--schedule-snapshot`? What is the idea behind this decision?

Author: good point 🤔 

I think there is no reason to persist env_var for theses options:
- `--import-snapshot`
- `--import-dump`
- `--ignore-missing-snapshot`
- `--ignore-snapshot-if-db-exists`

 no?

Reviewer: Mmmh, I see this more in a option file context. Where I would only have to add `--import-snapshot` when you need to quickly (or not) load a snapshot, and not all 5 options. But I also understand your point of view. @curquiza @eskombro @MarinPostma any input?

Reviewer 3: > I think there is no reason to persist env_var for theses options:

>    --import-snapshot
>    --import-dump
>    --ignore-missing-snapshot
>    --ignore-snapshot-if-db-exists

> no?

I don't know what to think. I have kind of the same feeling, but I'm not sure at all, I'm not used to creating devops files with a lot of env var...etc
I a

Reviewer 4: I think it's a good choice if two variables/options are meant to be used together, and you use just one, this shouldn't be just ignored without telling the user, and with this in mind, I agree with the requirement of `--import-snapshot` if you use `--ignore-missing-snapshot`. In the other hand, tehre are other cases where options are ignored, so just doing it here doesn't feel totally consistent.

I dont understand what you mean exactly @LegendreM by `I think there is no reason to persist env_var` so sorry if my answer is not pertnent.

Reviewer: Could we consider a warning instead of an error?

Reviewer 4: I like the warning idea 👍

Reviewer 5: I have been thinking about this. `--ignore-missing-snapshot` is an opt-in flag, I think that having an environement variable that sets this variable to always true is error prone. We should not design an API that leads our users to commit eventual error (it's very easy to forget that a variable in the environment has be set). Also, this flag makes not sense on it's own, and only makes sense at startup (unlike `--snapshot-dir`). The user shouldn't be allowed to launch a program with an invalid configuration (even if it appears like it does nothing bad). For this reason, I think this should be a hard error, and there should't be an env variable for this one, so previous configuration does not interfere with the current one.

Reviewer 4: Agree with @MarinPostma POV!

So IMO we should keep the hard error as he says, and remove the env var as @LegendreM proposed 👌

Author: removed `ENV_VAR` for:

-    --import-snapshot
-    --import-dump
-   --ignore-missing-snapshot
-    --ignore-snapshot-if-db-exists

Reviewer: What does this imply? I have a hard time understanding.

Reviewer: @LegendreM explained it to me. So now all our options have an env variable alternative except those 4? That is not really consistent...

Author: I think that `ENV_VAR` should not describe a starting behavior and should not be conflictual:
- if you both set `MEILI_IMPORT_DUMP` and `MEILI_IMPORT_SNAPSHOT`, it will conflict
- if you  set `MEILI_IMPORT_DUMP` you'll risk to delete data at the next start forgetting that you set this `ENV_VAR`

Reviewer: Oke, I understand. Apparently I was the one that did not understand what Marin said and everyone seems to agree. So lets go for that.

But this does not answer my original question: 
`where if you use --ignore-missing-snapshot without --import-snapshot it should not raise an error but a warning.`

Do we agree on this ?

Author: I don't like warnings

Reviewer: This discussion continues here: 
https://github.com/meilisearch/MeiliSearch/issues/1033

I tried to gather all opinions.

> ### Ignore Snapshot If DB Exists

Reviewer 2: ```suggestion
### Ignore Snapshot if DB Exists
```

**Environment variable**: `MEILI_IGNORE_SNAPSHOT_IF_DB_EXISTS`
**CLI option**: `--ignore-snapshot-if-db-exists`

> The engine skips snapshot importation if a database already exists. No error is thrown in this case.

Reviewer: ```suggestion
The engine skips snapshot importation if a database already exists, in the case of a `--snapshot-import`. No error is thrown in this case.
```

Author: ```suggestion
The engine skips snapshot importation if a database already exists, in the case of a `--import-snapshot`. No error is thrown in this case.
```

?

Reviewer: I want to make it clear that this variable has to do with the import snapshot process. You don't like it ?

> The engine skips snapshot importation if a database already exists. No error is thrown in this case.

Reviewer 2: ```suggestion
The engine skips importing the given snapshot if a database already exists, and instead loads the existing database. No error is thrown in this case.
```

Author: Sorry @react-learner , English is probably better but it's harder for me to understand your version. I needed to read it 4 times to understand it. Have you another sentence to submit?

Reviewer 2: @LegendreM You're right, it's confusing. How about this instead?

```suggestion
If a database already exists, MeiliSearch will attempt to launch using that database instead of importing a snapshot. No error is thrown in this case.
```
